src/fparser/two/tests/fortran2003/test_primary_r701.py

```diff
@@ -0,0 +1,196 @@
+            seen.append(subclass)
+            possible_subclasses(subclass, seen)
+    return seen
+
+
+def test_Constant(f2003_create):
+    '''Test that Constant types are matched by Primary.
+    '''
+    assert_subclass_parse(
+        '1.2e-03', f2003.Constant,
+        actual_type=f2003.Real_Literal_Constant,
+        expected_str='1.2E-03')
+
+
+def test_Designator(f2003_create):
+    '''Test that Designator types are matched by Primary.
+    '''
+    assert_subclass_parse(
+        'array(1:5)', f2003.Designator,
+        actual_type=f2003.Array_Section,
+        expected_str='array(1 : 5)')
+
+
+def test_Array_Constructor(f2003_create):
+    '''Test that Array Constructor types are matched by Primary.
+    '''
+    assert_subclass_parse(
+        '[ 1.2, 2.3e + 2,    -5.1 e-3 ]', f2003.Array_Constructor,
+        actual_type=f2003.Array_Constructor,
+        expected_str='[1.2, 2.3E+2, - 5.1E-3]')
+
+
+def test_Structure_Constructor(f2003_create):
+    '''Test that Structure Constructor types are matched by Primary.
+    '''
+    # Note: The actual returned type is Part_Ref. With more context of what
+    # has already been parsed it is possible that this could change in the
+    # future. For instance, in this example "PERSON" could actually be the
+    # name of a function.
+    assert_subclass_parse(
+        'PERSON ( 12,   "Jones" )', f2003.Structure_Constructor,
+        actual_type=f2003.Part_Ref,
+        expected_str='PERSON(12, "Jones")')
+
+
+@pytest.mark.xfail(reason="Unable to reach Function_Reference when parsing")
+def test_Function_Reference(f2003_create):
+    '''This test demonstrates the inability to distinguish
+    Structure_Constructor from Function_Reference without more parse context
+    than is currently being provided.
+    '''
+    assert_subclass_parse(
+        'a_function(1.2, some_kwarg="hello")', f2003.Function_Reference)
+
+
+@pytest.mark.xfail
+def test_Type_Param_Inquiry():
+    '''This test demonstrates the inability to distinguish Designator from
+    Type_Param_Inquiry without more parse context than is currently being
+    provided.
+    '''
+    assert_subclass_parse(
+        'X % KIND', f2003.Type_Param_Inquiry)
+
+
+def test_Type_Param_Name():
+    '''Test that Type_Param_Name types are matched by Primary.
+    '''
+    assert_subclass_parse(
+        'INTEGER', f2003.Type_Param_Name,
+        actual_type=f2003.Name,
+        expected_str='INTEGER')
+
+
+def test_Parenthesis(f2003_create):
+    '''Test that Parenthesis types are matched by Primary.
+    '''
+    assert_subclass_parse(
+        '(a +  b)', f2003.Parenthesis,
+        expected_str='(a + b)')
+
+
+def test_no_match(f2003_create):
+    '''Test that a NoMatchError is raised if we provide code
+    that isn't allowed as a Primary type (e.g. a comment).
+    '''
+    with pytest.raises(fparser.two.utils.NoMatchError):
+        obj = f2003.Primary('! A comment')
+
+
+@pytest.mark.xfail
```

Comment:
Up until now we've not added xfail tests for constraints we can't check. I can see the advantage of doing this though so would be happy to add it as a rule from now on. The only thing I would ask is that you give a reason and reference an issue (it can be the same as before).